src/main/java/org/jitsi/videobridge/cc/vp8/VP8FrameProjection.java

```diff
@@ -0,0 +1,461 @@
+    }
+
+    /**
+     * Small utility method that computes and returns the starting sequence
+     * number to use in the projection of the frame that will be sent
+     * immediately after the {@link #vp8Frame}.
+     *
+     * @return the starting sequence number to use in the projection of the
+     * frame that will be sent immediately after the {@link #vp8Frame}.
+     */
+    private int nextStartingSequenceNumber()
+    {
+        int vp8FrameLength = RTPUtils.getSequenceNumberDelta(
+            vp8Frame.getMaxSequenceNumber(),
+            vp8Frame.getStartingSequenceNumber());
+
+        int nextStartingSequenceNumber
+            = startingSequenceNumber + vp8FrameLength + 1;
+
+        return nextStartingSequenceNumber & RawPacket.SEQUENCE_NUMBER_MASK;
+    }
+
+    /**
+     * Small utility method that computes and returns the extended picture id to
+     * use in the projection of the frame that will be sent immediately after
+     * the {@link #vp8Frame}.
+     *
+     * @return the extended picture id to use in the projection of the frame
+     * that will be sent immediately after the {@link #vp8Frame}.
+     */
+    private int nextExtendedPictureId()
+    {
+        return (extendedPictureId + 1)
+            & DePacketizer.VP8PayloadDescriptor.EXTENDED_PICTURE_ID_MASK;
+    }
+
+    /**
+     * Rewrites an RTP packet that belongs to {@link #vp8Frame}. If this is the
+     * first packet of a frame and if a packet cache has been specified, we
+     * piggy-backing any missed packets.
+     *
+     * @param cache the cache to pull piggy-backed packets from.
+     * @param rtpPacket the RTP packet to project.
+     */
+    RawPacket[] rewriteRtp(@NotNull RawPacket rtpPacket, RawPacketCache cache)
+    {
+        int originalSequenceNumber = rtpPacket.getSequenceNumber();
+
+        rewriteRtpInternal(rtpPacket);
+
+        if (originalSequenceNumber != vp8Frame.getStartingSequenceNumber())
+        {
+            return AdaptiveTrackProjection.EMPTY_PACKET_ARR;
+        }
+
+        // We piggy-back any re-ordered packets of this frame.
+        long vp8FrameSSRC = vp8Frame.getSSRCAsLong()
+            , vp8FrameTimestamp = vp8Frame.getTimestamp();
+
+        List<RawPacket> piggyBackedPackets = new ArrayList<>();
+        boolean done = cache == null;
+        while (!done)
+        {
+            int piggyBackedPacketSequenceNumber
+                = (originalSequenceNumber + 1 + piggyBackedPackets.size())
+                & RawPacket.SEQUENCE_NUMBER_MASK;
+
+            RawPacket lastPacket = cache.get(
+                vp8FrameSSRC, piggyBackedPacketSequenceNumber);
+
+            if (lastPacket == null
+                || vp8FrameTimestamp != lastPacket.getTimestamp())
+            {
+                done = true;
+            }
+            else
+            {
+                // the call to accept (synchronized) may update the
+                // maxSequenceNumber.
+                if (accept(lastPacket))
```

Comment:
is it weird to have this method call accept on its own?  i assume this accept is normally called by a layer above?  should that be done by a layer above this after all the piggy-backed packets are returned?  we don't seem to access `maxSequenceNumber` here, so wondering why we need to call this to update it

Reply:
This happens in the transform thread, so by the time we get to rewrite the definitively accepted first packet of a frame, the first packet of another frame may have already been accepted, which means there's no longer space to piggyback anything.